File: ports/1.21.1-full/src/main/java/mods/eln/node/transparent/TransparentNodeItem.java
package mods.eln.node.transparent;

import mods.eln.generic.GenericItemBlockUsingDamage;
import mods.eln.ghost.GhostGroup;
import mods.eln.misc.Coordinate;
import mods.eln.misc.Direction;
import mods.eln.misc.Utils;
import mods.eln.node.NodeBlock;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.InteractionHand;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;

public class TransparentNodeItem extends GenericItemBlockUsingDamage<TransparentNodeDescriptor> {


    public TransparentNodeItem(Block b) {
        super(b);
        setHasSubtypes(true);
    }


    @Override
    public InteractionResult onItemUse(Player player, Level world, BlockPos pos, InteractionHand hand, net.minecraft.core.Direction facing, float hitX, float hitY, float hitZ) {
        ItemStack stack = player.getHeldItem(hand);
        if (stack.isEmpty()) return InteractionResult.FAIL;

        BlockState iblockstate = world.getBlockState(pos);
        Block block = iblockstate.getBlock();

        if (!block.isReplaceable(world, pos)) {
            pos = pos.offset(facing);
        }

        if (player.canPlayerEdit(pos, facing, stack) && world.mayPlace(this.block, pos, false, facing, null)) {
            TransparentNodeDescriptor descriptor = getDescriptor(stack);
            if (descriptor == null) return InteractionResult.FAIL;

            Direction direction = Direction.fromFacing(facing).getInverse();
            Direction front = descriptor.getFrontFromPlace(direction, player);

            // Apply spawn delta
            int[] v = new int[]{descriptor.getSpawnDeltaX(), descriptor.getSpawnDeltaY(), descriptor.getSpawnDeltaZ()};
            front.rotateFromXN(v);
            BlockPos adjustedPos = pos.add(v[0], v[1], v[2]);

            if (!world.getBlockState(adjustedPos).getBlock().isReplaceable(world, adjustedPos)) {
                return InteractionResult.FAIL;
            }

            Coordinate coord = new Coordinate(adjustedPos, world);
            String error = descriptor.checkCanPlace(coord, front);
            if (error != null) {
                if (!world.isRemote) Utils.sendMessage(player, error);
                return InteractionResult.FAIL;
            }

            if (world.isRemote) return InteractionResult.SUCCESS;

            // Plot ghosts
            GhostGroup ghostgroup = descriptor.getGhostGroup(front);
            if (ghostgroup != null) ghostgroup.plot(coord, coord, descriptor.getGhostGroupUuid());

            // Create Node
            TransparentNode node = new TransparentNode();
            node.onBlockPlacedBy(coord, front, player, stack);

            // Set block state
            int metadata = node.getBlockMetadata();
            BlockState newState = this.block.getStateFromMeta(metadata);
            if (world.setBlockState(adjustedPos, newState, 3)) {
                // Play placement sound
                SoundType soundtype = this.block.getSoundType(newState, world, adjustedPos, player);
                world.playSound(player, adjustedPos, soundtype.getPlaceSound(), SoundSource.BLOCKS, (soundtype.getVolume() + 1.0F) / 2.0F, soundtype.getPitch() * 0.8F);
                
                // Notify block
                ((NodeBlock) this.block).onBlockPlacedBy(world, adjustedPos, front, player, newState);
                
                stack.shrink(1);
                node.checkCanStay(true);
                return InteractionResult.SUCCESS;
            }
        }

        return InteractionResult.FAIL;
    }

    @Override
    public boolean placeBlockAt(ItemStack stack, Player player, Level world, BlockPos pos, net.minecraft.core.Direction side, float hitX, float hitY, float hitZ, BlockState state) {
        // Handled in onItemUse
        return false;
    }

    // TODO(1.10): Fix item rendering.
//    @Override
//    public boolean handleRenderType(ItemStack item, ItemRenderType type) {
//        TransparentNodeDescriptor d = getDescriptor(item);
//        if (Utils.nullCheck(d)) return false;
//        return d.handleRenderType(item, type);
//    }
//
//    @Override
//    public boolean shouldUseRenderHelper(ItemRenderType type, ItemStack item,
//                                         ItemRendererHelper helper) {
//
//        return getDescriptor(item).shouldUseRenderHelper(type, item, helper);
//    }
//
//    public boolean shouldUseRenderHelperEln(ItemRenderType type, ItemStack item, ItemRendererHelper helper) {
//        return getDescriptor(item).shouldUseRenderHelperEln(type, item, helper);
//    }
//
//    @Override
//    public void renderItem(ItemRenderType type, ItemStack item, Object... data) {
//        Minecraft.getMinecraft().profiler.startSection("TransparentNodeItem");
//
//        if (shouldUseRenderHelperEln(type, item, null)) {
//            switch (type) {
//                case ENTITY:
//                    GL11.glTranslatef(0.00f, 0.3f, 0.0f);
//                    break;
//                case EQUIPPED_FIRST_PERSON:
//                    GL11.glTranslatef(0.50f, 1, 0.5f);
//                    break;
//                case EQUIPPED:
//                    GL11.glTranslatef(0.50f, 1, 0.5f);
//                    break;
//                case FIRST_PERSON_MAP:
//                    break;
//                case INVENTORY:
//                    GL11.glRotatef(90, 0, 1, 0);
//                    break;
//                default:
//                    break;
//            }
//        }
//        getDescriptor(item).renderItem(type, item, data);
//
//        Minecraft.getMinecraft().profiler.endSection();
//    }
}

File: src/main/kotlin/mods/eln/node/transparent/TransparentNodeItem.kt
@file:Suppress("NAME_SHADOWING")
package mods.eln.node.transparent

import mods.eln.generic.DescriptorBlockItem
import mods.eln.generic.GenericItemBlockUsingDamage
import net.minecraft.world.item.Item
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.item.context.UseOnContext
import net.minecraft.world.InteractionResult
import net.minecraft.sounds.SoundSource
import java.util.function.Supplier
import mods.eln.misc.Coordinate
import mods.eln.misc.Direction.Companion.fromIntMinecraftSide
import mods.eln.misc.Utils.sendMessage
import mods.eln.misc.Utils.nullCheck
import mods.eln.node.NodeBlock
import net.minecraft.world.level.block.Block
import net.minecraft.client.Minecraft
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import mods.eln.client.itemrender.IItemRenderer
import mods.eln.client.itemrender.IItemRenderer.ItemRenderType
import mods.eln.client.itemrender.IItemRenderer.ItemRendererHelper
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.core.Direction as EnumFacing
import net.minecraft.core.BlockPos
import mods.eln.client.gl.GL11
import mods.eln.misc.setBlock

/**
 * The transparent-node element family: one [Placer] per descriptor (the Flattening); the block
 * goes to the descriptor's spawn offset from the clicked cell, so vanilla's placement is replaced
 * rather than adapted (1.7.10's `placeBlockAt`).
 */
class TransparentNodeItem(b: Supplier<Block>) : GenericItemBlockUsingDamage<TransparentNodeDescriptor?>(b, "TransparentNodeItem"), IItemRenderer {

    class Placer(family: TransparentNodeItem, descriptor: TransparentNodeDescriptor, id: Int, block: Block, properties: Properties) :
        DescriptorBlockItem<TransparentNodeDescriptor?>(family, descriptor, id, block, properties) {
        override fun useOn(context: UseOnContext): InteractionResult {
            val player = context.player ?: return InteractionResult.PASS
            val place = BlockPlaceContext(context)
            if (!place.canPlace()) return InteractionResult.FAIL
            val stack = context.itemInHand
            if (!player.mayUseItemAt(place.clickedPos, context.clickedFace, stack)) return InteractionResult.FAIL
            if (context.level.isClientSide) return InteractionResult.SUCCESS
            if (!(family as TransparentNodeItem).placeBlockAt(stack, player, context.level, place.clickedPos, context.clickedFace)) return InteractionResult.FAIL
            val sound = context.level.getBlockState(place.clickedPos).getSoundType(context.level, place.clickedPos, player)
            context.level.playSound(null, place.clickedPos, sound.placeSound, SoundSource.BLOCKS, (sound.volume + 1.0f) / 2.0f, sound.pitch * 0.8f)
            stack.shrink(1)
            return InteractionResult.SUCCESS
        }
    }

    override fun newItem(id: Int, descriptor: TransparentNodeDescriptor?): Item =
        Placer(this, descriptor!!, id, block.get(), newProperties(descriptor))

    /** Server side: node first, then the block at the descriptor's spawn offset. */
    fun placeBlockAt(stack: ItemStack, player: Player, world: Level, pos: BlockPos, side: EnumFacing): Boolean {
        var x = pos.x
        var y = pos.y
        var z = pos.z
        if (world.isClientSide) return false
        val descriptor = getDescriptor(stack) ?: return false
        val direction = fromIntMinecraftSide(side.get3DDataValue())!!.inverse
        val front = descriptor.getFrontFromPlace(direction, player)
        val v = intArrayOf(descriptor.spawnDeltaX, descriptor.spawnDeltaY, descriptor.spawnDeltaZ)
        front!!.rotateFromXN(v)
        x += v[0]
        y += v[1]
        z += v[2]
        val coord = Coordinate(x, y, z, world)
        var error: String?
        if (descriptor.checkCanPlace(coord, front).also { error = it } != null) {
            sendMessage(player, error)
            return false
        }
        val ghostgroup = descriptor.getGhostGroupFront(front)
        ghostgroup?.plot(coord, coord, descriptor.ghostGroupUuid)
        val node = TransparentNode()
        node.onBlockPlacedBy(coord, front, player, stack)
        val block = this.block.get() as NodeBlock
        world.setBlock(x, y, z, block, node.blockMetadata, 0x03) //caca1.5.1
        block.onBlockPlacedBy(world, BlockPos(x, y, z), direction, player, node.blockMetadata)
        node.checkCanStay(true)
        return true
    }

    override fun handleRenderType(item: ItemStack, type: ItemRenderType): Boolean {
        val d = getDescriptor(item)
        return if (nullCheck(d)) false else d!!.handleRenderType(item, type)
    }

    override fun shouldUseRenderHelper(type: ItemRenderType, item: ItemStack,
                                       helper: ItemRendererHelper): Boolean {
        val descriptor = getDescriptor(item) ?: return false
        return descriptor.shouldUseRenderHelper(type, item, helper)
    }

    fun shouldUseRenderHelperEln(type: ItemRenderType?, item: ItemStack?, helper: ItemRendererHelper?): Boolean {
        val descriptor = getDescriptor(item) ?: return false
        return descriptor.shouldUseRenderHelperEln(type, item, helper)
    }

    override fun renderItem(type: ItemRenderType, item: ItemStack, vararg data: Any) {
        Minecraft.getInstance().profiler.push("TransparentNodeItem")
        if (shouldUseRenderHelperEln(type, item, null)) {
            when (type) {
                ItemRenderType.ENTITY -> GL11.glTranslatef(0.00f, 0.3f, 0.0f)
                ItemRenderType.EQUIPPED_FIRST_PERSON -> GL11.glTranslatef(0.50f, 1f, 0.5f)
                ItemRenderType.EQUIPPED -> GL11.glTranslatef(0.50f, 1f, 0.5f)
                ItemRenderType.FIRST_PERSON_MAP -> {
                }
                ItemRenderType.INVENTORY -> GL11.glRotatef(90f, 0f, 1f, 0f)
                else -> {
                }
            }
        }
        val descriptor = getDescriptor(item)
        if (descriptor != null) {
            descriptor.renderItem(type, item, *data)
        }
        Minecraft.getInstance().profiler.pop()
    }

}
